### ISSUE TITLE
java.lang.ClassCastException

### ISSUE THREAD

**Issue by duygithub** [versly/wsdoc#16]
Greetings,

I'm trying to setup wsdoc working on our maven builds.  I've made 3 changes to the original pom configuration:

1) The incorrect markup of arguments has been fixed to:

``` XML
<arguments>
  <argument>${project.build.directory}/classes</argument>
</arguments>
```

2) I've had to exclude the joda-time because we use a newer version in our production code

``` XML
<dependency>
  <groupId>org.versly</groupId>
  <artifactId>versly-wsdoc</artifactId>
  <version>1.0-SNAPSHOT</version>
  <scope>compile</scope>
  <exclusions>
    <exclusion>
      <groupId>joda-time</groupId>
      <artifactId>joda-time</artifactId>
    </exclusion>
  </exclusions>
</dependency>
```

3) I've updated to a newer version of maven-processor-plugin (1.3.6 didn't work for me either)

``` XML
<groupId>org.bsc.maven</groupId>
<artifactId>maven-processor-plugin</artifactId>
<version>2.2.4</version>
```

I ran into the following error when I ran the following command:
_mvn clean install -e -DskipTests=true_

``` Java
[ERROR] Failed to execute goal org.bsc.maven:maven-processor-plugin:2.2.4:process (default) on project ws-api: Error executing: java.lang.ClassCastException: com.sun.tools.javac.code.Type$WildcardType cannot be cast to javax.lang.model.type.TypeVariable -> [Help 1]
```

The full-stack trace is as follows:

``` Java
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.bsc.maven:maven-processor-plugin:2.2.4:process (default) on project ws-api: Error executing
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
        at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
        at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
        at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
        at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
        at org.apache.maven.cli.MavenCli.execute(MavenCli.java:537)
        at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:196)
        at org.apache.maven.cli.MavenCli.main(MavenCli.java:141)
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:57)
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.lang.reflect.Method.invoke(Method.java:601)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:290)
        at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:230)
        at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:409)
        at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:352)
Caused by: org.apache.maven.plugin.MojoExecutionException: Error executing
        at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.execute(AbstractAnnotationProcessorMojo.java:292)
        at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
        at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
        ... 19 more
Caused by: java.lang.RuntimeException: java.lang.ClassCastException: com.sun.tools.javac.code.Type$WildcardType cannot be cast to javax.lang.model.type.TypeVariable
        at com.sun.tools.javac.main.Main.compile(Main.java:469)
        at com.sun.tools.javac.api.JavacTaskImpl.call(JavacTaskImpl.java:132)
        at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.executeWithExceptionsHandled(AbstractAnnotationProcessorMojo.java:545)
        at org.bsc.maven.plugin.processor.AbstractAnnotationProcessorMojo.execute(AbstractAnnotationProcessorMojo.java:285)
        ... 21 more
Caused by: java.lang.ClassCastException: com.sun.tools.javac.code.Type$WildcardType cannot be cast to javax.lang.model.type.TypeVariable
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.recurseForJsonType(AnnotationProcessor.java:481)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.addFieldFromBeanMethod(AnnotationProcessor.java:464)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:440)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildType(AnnotationProcessor.java:399)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:384)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:286)
        at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:731)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeForDeclaredType(AnnotationProcessor.java:256)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeFromTypeMirror(AnnotationProcessor.java:232)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeFromTypeMirror(AnnotationProcessor.java:237)
        at org.versly.rest.wsdoc.AnnotationProcessor.access$000(AnnotationProcessor.java:52)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.addFieldFromBeanMethod(AnnotationProcessor.java:468)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:440)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildType(AnnotationProcessor.java:399)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:384)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:286)
        at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:731)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeForDeclaredType(AnnotationProcessor.java:256)
        at org.versly.rest.wsdoc.AnnotationProcessor.access$300(AnnotationProcessor.java:52)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.recurseForJsonType(AnnotationProcessor.java:485)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.addFieldFromBeanMethod(AnnotationProcessor.java:464)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:440)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildType(AnnotationProcessor.java:399)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:384)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:286)
        at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:731)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeForDeclaredType(AnnotationProcessor.java:256)
        at org.versly.rest.wsdoc.AnnotationProcessor.access$300(AnnotationProcessor.java:52)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.recurseForJsonType(AnnotationProcessor.java:485)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.addFieldFromBeanMethod(AnnotationProcessor.java:464)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:440)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:435)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildType(AnnotationProcessor.java:399)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:384)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:286)
        at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:731)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeForDeclaredType(AnnotationProcessor.java:256)
        at org.versly.rest.wsdoc.AnnotationProcessor.access$300(AnnotationProcessor.java:52)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.recurseForJsonType(AnnotationProcessor.java:485)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.addFieldFromBeanMethod(AnnotationProcessor.java:464)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildTypeContents(AnnotationProcessor.java:440)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.buildType(AnnotationProcessor.java:399)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:384)
        at org.versly.rest.wsdoc.AnnotationProcessor$TypeVisitorImpl.visitDeclared(AnnotationProcessor.java:286)
        at com.sun.tools.javac.code.Type$ClassType.accept(Type.java:731)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeForDeclaredType(AnnotationProcessor.java:256)
        at org.versly.rest.wsdoc.AnnotationProcessor.jsonTypeFromTypeMirror(AnnotationProcessor.java:232)
        at org.versly.rest.wsdoc.AnnotationProcessor.buildResponseFormat(AnnotationProcessor.java:268)
        at org.versly.rest.wsdoc.AnnotationProcessor.processRequestMappingMethod(AnnotationProcessor.java:136)
        at org.versly.rest.wsdoc.AnnotationProcessor.processElements(AnnotationProcessor.java:103)
        at org.versly.rest.wsdoc.AnnotationProcessor.process(AnnotationProcessor.java:67)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.callProcessor(JavacProcessingEnvironment.java:793)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.discoverAndRunProcs(JavacProcessingEnvironment.java:722)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.access$1700(JavacProcessingEnvironment.java:97)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment$Round.run(JavacProcessingEnvironment.java:1029)
        at com.sun.tools.javac.processing.JavacProcessingEnvironment.doProcessing(JavacProcessingEnvironment.java:1163)
        at com.sun.tools.javac.main.JavaCompiler.processAnnotations(JavaCompiler.java:1108)
        at com.sun.tools.javac.main.JavaCompiler.compile(JavaCompiler.java:824)
        at com.sun.tools.javac.main.Main.compile(Main.java:439)
        ... 24 more
```

My plugin config is as follows:

``` XML
<build>
  <plugins>
    <!-- REST documentation parse -->
    <plugin>
      <groupId>org.bsc.maven</groupId>
      <artifactId>maven-processor-plugin</artifactId>
      <version>2.2.4</version>

      <configuration>
        <outputDiagnostics>true</outputDiagnostics>
        <processors>
          <processor>org.versly.rest.wsdoc.AnnotationProcessor</processor>
        </processors>
      </configuration>

      <executions>
        <execution>
          <phase>compile</phase>
          <goals>
            <goal>process</goal>
          </goals>
        </execution>
      </executions>

      <dependencies>
        <dependency>
          <groupId>org.versly</groupId>
          <artifactId>versly-wsdoc</artifactId>
          <version>1.0-SNAPSHOT</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <groupId>joda-time</groupId>
              <artifactId>joda-time</artifactId>
            </exclusion>
          </exclusions>
        </dependency>
      </dependencies>
    </plugin>


    <!-- REST HTML documentation generation. Possibly in a different POM
           than the parse step above. -->
    <plugin>
      <groupId>org.codehaus.mojo</groupId>
      <artifactId>exec-maven-plugin</artifactId>
      <version>1.2</version>

      <executions>
        <execution>
          <phase>prepare-package</phase>
          <goals>
            <goal>java</goal>
          </goals>
          <configuration>
            <mainClass>org.versly.rest.wsdoc.RestDocAssembler</mainClass>
            <arguments>
              <argument>${project.build.directory}/classes</argument>
            </arguments>
          </configuration>
        </execution>
      </executions>

      <dependencies>
        <dependency>
          <groupId>org.versly</groupId>
          <artifactId>versly-wsdoc</artifactId>
          <version>1.0-SNAPSHOT</version>
          <scope>compile</scope>
          <exclusions>
            <exclusion>
              <groupId>joda-time</groupId>
              <artifactId>joda-time</artifactId>
            </exclusion>
          </exclusions>
        </dependency>
      </dependencies>
    </plugin>

  </plugins>
</build>
```

Any help would be greatly appreciated.

Thank you.

**Comment by kjellmrud** [comment]
This issue has hit me as well. It happens on line 663 of AnnotationProcessor.java:

      concreteTypes.add(_typeArguments.get(((TypeVariable) generic).asElement().getSimpleName()));

The variable 'generic' fails casting to javax.lang.model.type.TypeVariable as it is of type javax.lang.model.type.WildcardType. This happens to me when I use a return from my endpoint of type WebAsyncTask<ResponseEntity<MyResponseObject>>.

**Comment by pcl** [comment]
Resolved in e6e5a12.
